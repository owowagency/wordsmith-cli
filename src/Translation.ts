


export class Translation {

    private readonly _key: string;

    private readonly _values: TranslationValue[];

    constructor(responseObject: any) {
        this._key = responseObject.key;
        this._values = responseObject.translationValues.map((translation: any) => {
            return new TranslationValue(translation);
        });
    }

    get values(): TranslationValue[] {
        return this._values;
    }

    get key(): string {
        return this._key;
    }
}

class TranslationValue {

    language: string;
    value: string;

    constructor(responseObject: any) {
        this.language = responseObject.language;
        this.value = responseObject.value;
    }
}