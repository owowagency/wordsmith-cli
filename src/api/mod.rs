use log::{error, log_enabled, debug};
use reqwest::{ClientBuilder, header::{HeaderMap, HeaderValue, InvalidHeaderValue}, Client, StatusCode, Request, Response};
use serde::de::DeserializeOwned;
use thiserror::Error;

use crate::environment::AccessToken;

use self::models::error::ApiError;

pub mod models;
pub mod pull;
pub mod push;
pub mod info;

const BASE_URL: &'static str = env!("BASE_URL");
const USER_AGENT: &'static str = env!("USER_AGENT");

pub struct WordsmithClient {
    headers: HeaderMap,
    client: Client,
}

type Result<T> = std::result::Result<T, WordsmithError>;

#[derive(Debug, Error)]
pub enum WordsmithError {
    #[error("Unable to create API Client, reason: {0}")]
    Init(String),
    #[error("HTTP Client error, reason: {0}")]
    Http(String),
    #[error("API Error, status: {0}, message: {1}")]
    Api(StatusCode, ApiError),
    #[error("IO Error, reason: {0}")]
    Io(String),
}

impl From<InvalidHeaderValue> for WordsmithError {
    fn from(value: InvalidHeaderValue) -> Self {
        WordsmithError::Init(value.to_string())
    }
}

impl From<reqwest::Error> for WordsmithError {
    fn from(value: reqwest::Error) -> Self {
        let status = value.status();
        let url = value.url();
        let message = format!("HTTP error [status: {:?}, url: {:?}], message: {}", status, url, value.to_string());
        WordsmithError::Http(message)
    }
}

impl WordsmithClient {
    pub fn new(token: Option<&AccessToken>) -> Result<Self> {
        let user_agent = HeaderValue::from_str(USER_AGENT)?;
        let accept = HeaderValue::from_str("application/json")?;
        let mut headers = HeaderMap::new();
        headers.append("User-Agent", user_agent);
        headers.append("Accept", accept);
        if let Some(token) = token {
            let raw_token = token.get_token().map_err(|op| WordsmithError::Init(op))?;
            let token_header = HeaderValue::from_str(&format!("Bearer {}", raw_token))?;
            headers.append("Authorization", token_header);
        }
        let client = ClientBuilder::new()
            .default_headers(headers.clone())
            .build()
            .map_err(|op| WordsmithError::Init(op.to_string()))?;
        
        Ok(Self {
            client,
            headers,
        })
    }

    async fn send(&self, request: Request) -> Result<Response> {
        if log_enabled!(log::Level::Debug) {
            let url = request.url();
            let query = url.query_pairs();
            let headers = request.headers();
            let mut buffer = String::new();
            buffer.push_str(&format!("{} {}", request.method(), request.url()));
            for (name, value) in self.headers.iter() {
                buffer.push_str(&format!("\n{}: {}", name.as_str(), value.to_str().ok().unwrap_or("")))
            }
            for (name, value) in headers.iter() {
                buffer.push_str(&format!("\n{}: {}", name.as_str(), value.to_str().ok().unwrap_or("")))
            }
            for (key, value) in query.into_iter() {
                buffer.push_str(&format!("\n{}={}", key, value))
            }
            debug!("{}", buffer);
        }
        return Ok(self.client.execute(request).await?);
    }

    pub async fn execute<Res : DeserializeOwned>(&self, request: Request) -> Result<Res> {
        let response = self.send(request).await?;

        let status = response.status();

        match status.as_u16() {
            200..=299 => {
                let response: Res = response.json().await?;
                Ok(response)
            },
            _ => Err(WordsmithClient::decode_error_response(response).await),
        }
    }

    pub async fn execute_no_content(&self, request: Request) -> Result<()> {
        let response = self.send(request).await?;
        let status = response.status();

        if status != StatusCode::NO_CONTENT {
            return Err(WordsmithClient::decode_error_response(response).await);
        }

        Ok(())
    }

    pub async fn download(&self, url: &str) -> Result<Vec<u8>> {
        let request = self.client.get(url).build()?;
        let response = self.send(request).await?;
        let status = response.status();

        if status != StatusCode::OK {
            return Err(WordsmithClient::decode_error_response(response).await);
        }

        return Ok(response.bytes().await?.to_vec());
    }

    async fn decode_error_response(response: Response) -> WordsmithError {
        let status = response.status();
        let content_type = match response.headers().get("Content-Type") {
            Some(header) => header.to_str().unwrap_or(""),
            None => "",
        };

        if content_type.contains("application/json") {
            return match response.json().await {
                Ok(error) => WordsmithError::Api(status, error),
                Err(err) => WordsmithError::Api(status, ApiError { 
                    message: err.to_string(),
                    errors: vec![],
                })
            };
        }

        match response.text().await {
            Ok(text) => WordsmithError::Api(status, ApiError {
                message: text,
                errors: vec![],
            }),
            Err(err) => err.into(),
        } 
    }
}

