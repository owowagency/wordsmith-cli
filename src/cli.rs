use clap::{command, Parser, Subcommand, Args};

use crate::environment::Environment;

#[derive(Parser)]
#[command(author, version, about, long_about = None)]
pub struct CommandLine {
    #[command(subcommand)]
    pub command: Command,
}

#[derive(Subcommand)] 
pub enum Command {
    Pull(PullArgs),
    Push(PushArgs),
}

#[derive(Debug, Args)]
pub struct GlobalArgs {
    #[arg(short, long, default_value = "wordsmith.yml", value_parser = clap::value_parser!(Environment))]
    pub env: Environment,
    /// Enable verbose logging
    #[arg(long, default_value_t = false)]
    pub verbose: bool
}

#[derive(Debug, Args)]
pub struct PushArgs {
    #[clap(flatten)]
    pub global: GlobalArgs,
    /// Overwrites existing strings
    #[arg(short = 'f', long = "force", default_value_t = false)]
    pub overwrite: bool,
    /// Verify translations
    #[arg(short, long, default_value_t = false)]
    pub verify: bool,
    #[arg(short = 'd', long = "dry-run", default_value_t = false)]
    pub dry_run: bool,
}

#[derive(Debug, Args)]
pub struct PullArgs {
    #[clap(flatten)]
    pub global: GlobalArgs,
    #[arg(short = 'd', long = "dry-run", default_value_t = false)]
    pub dry_run: bool,
}