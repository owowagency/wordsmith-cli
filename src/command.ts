export const test = () => {
  console.log('yoooo');
}
