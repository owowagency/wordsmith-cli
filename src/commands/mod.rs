use async_trait::async_trait;

use crate::{cli::Command, api::WordsmithError};

mod list_projects;
mod pull;
mod push;
mod helpers;

#[async_trait]
pub trait Execute {
    async fn execute(&self) -> Result<(), WordsmithError>;
}

#[async_trait]
impl Execute for Command {
    async fn execute(&self) -> Result<(), WordsmithError> {
        match self {
            Command::ListProjects(args) => args.execute().await?,
            Command::Pull(args) => args.execute().await?,
            Command::Push(args) => args.execute().await?,
        };

        Ok(())
    }
}
