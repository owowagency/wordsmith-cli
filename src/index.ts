const { program } = require("commander");
const packageJson = require('../package.json');

import { Config } from './config';

let config = new Config();

program
    .name(config.get('cli.name'))
    .description(packageJson.description)
    .version(packageJson.version)
    .option('-token, --api-token <token>', 'Set the token used to authenticate with the API')
    .option('-id, --project-id <id>', 'Set the ID of the project that will be used for translations')
    .option('-d, --translations-directory <directory>', 'Set the directory where the translations will be stored')
    .option('-t --tags <tags>', 'Set the tags that will be used to filter the translations');

program.parse(process.argv);

const options = program.opts();

if (options.apiToken) { config.set('client.api_token', options.apiToken); }

if (options.projectId) { config.set('client.project_id', options.projectId); }

if (options.translationsDirectory) { config.set('client.translations_directory', options.translationsDirectory); }

if (options.tags) { config.set('client.tags', options.tags); }

if (! config.isConfigured()) {
    console.log("Please configure the CLI before using it.");
    program.help();
}