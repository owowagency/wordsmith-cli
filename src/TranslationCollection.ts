import {Translation} from "./Translation";

export class TranslationCollection {

    private translations: Translation[];

    constructor(translationResponse: any) {
        this.translations = translationResponse.map((translation: any) => {
            return new Translation(translation);
        });
    }

    uniqueLanguages(): string[] {
        let languageSet: Set<string> = new Set();

        this.translations.forEach((translation: Translation) => {
            translation.values.forEach(v => languageSet.add(v.language))
        });

        return Array.from(languageSet);
    }

    getTranslationsForLanguage(language: string): Translation[] {
        return this.translations.filter((translation: Translation) => {
            return translation.values.some(v => v.language === language);
        });
    }

    toJson(language: string): any {
        return this.getTranslationsForLanguage(language)
            .reduce((translations: Record<string, any>, translation: Translation) => {
                translations[translation.key] = translation.values.find(v => v.language === language)?.value;

                return translations;
            }, {}
        );
    }
}