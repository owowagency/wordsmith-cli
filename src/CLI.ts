import {Repository} from "./Repository";
import {Config} from "./Config";
import {Storage} from "./Storage";
import {StorageTranslationCollection} from "./StorageTranslationCollection";
import fs from "fs";
import {TranslationCollection} from "./TranslationCollection";

/**
 * A class that contains all functionality of the CLI.
 */
export class CLI {

    /**
     * Reads translations from remote repository.
     */
    static async pull(repository: Repository, config: Config): Promise<void> {
        if (! config.isPathValid()) {
            try {
                fs.mkdirSync(config.get('translations_directory'), { recursive: true })
            } catch (error) {
                throw new Error("Invalid translations directory: " + config.get('translations_directory'));
            }
        }

        let rawProjectTranslations: Object[] = await repository.getRawProjectTranslations();

        let translationCollection: TranslationCollection = new TranslationCollection(rawProjectTranslations);

        let storage: Storage = new Storage(config.get('translations_directory'));

        storage.writeTranslationsForAllLanguages(translationCollection);
    }

    /**
     * Stores translations in remote repository.
     */
    static async push(repository: Repository, config: Config): Promise<void> {
        if (! config.isPathValid()) {
            throw new Error("Invalid translations directory: " + config.get('translations_directory'));
        }

        let storage: Storage = new Storage(config.get('translations_directory'));

        let storageTranslationCollection: StorageTranslationCollection = storage.readTranslations();

        let translationsJson: any = storageTranslationCollection.toJson();

        let success: boolean = await repository.sendRawProjectTranslations(translationsJson);

        if (! success) {
            throw new Error("Error sending translations to the server.");
        }
    }
}