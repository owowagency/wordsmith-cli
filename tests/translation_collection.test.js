import {Translation} from "../src/Translation";
import expect from "expect";
import {TranslationCollection} from "../src/TranslationCollection";

const dutchLithuanianTranslation = {
    "key": "promiscuous girl",
    "tags": [],
    "translationValues": [
        {
            "language": "NL",
            "value": "promiscue meisje"
        },
        {
            "language": "LT",
            "value": "pasileidusi mergina"
        }
    ]
}

const dutchTranslation = {
    "key": "bladee",
    "tags": [],
    "translationValues": [
        {
            "language": "NL",
            "value": "Waylon"
        },
    ]
}

const responseObject = [dutchLithuanianTranslation, dutchTranslation];

test('constructor creates translation collection from a response object', () => {

    let translationCollection = new TranslationCollection(responseObject);

    expect(translationCollection).toMatchSnapshot();
});

test('unique languages returns only unique languages', () => {

    let translationCollection = new TranslationCollection(responseObject);

    expect(translationCollection.uniqueLanguages()).toEqual(["NL", "LT"]);
});

test('getTranslationsForLanguage returns only translations for a given language', () => {

        let translationCollection = new TranslationCollection(responseObject);

        let lithuanianTranslations = translationCollection.getTranslationsForLanguage("LT");

        expect(lithuanianTranslations).toEqual([new Translation(dutchLithuanianTranslation)]);
});

test('toJson returns json with translations for each language', () => {

        let translationCollection = new TranslationCollection(responseObject);

        let languages = translationCollection.uniqueLanguages();

        languages.forEach(language => {
            expect(translationCollection.toJson(language)).toMatchSnapshot();
        });
});