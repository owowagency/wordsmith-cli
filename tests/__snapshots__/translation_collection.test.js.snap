// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor creates translation collection from a response object 1`] = `
TranslationCollection {
  "translations": [
    Translation {
      "_key": "promiscuous girl",
      "_values": [
        TranslationValue {
          "language": "NL",
          "value": "promiscue meisje",
        },
        TranslationValue {
          "language": "LT",
          "value": "pasileidusi mergina",
        },
      ],
    },
    Translation {
      "_key": "bladee",
      "_values": [
        TranslationValue {
          "language": "NL",
          "value": "Waylon",
        },
      ],
    },
  ],
}
`;

exports[`toJson returns json with translations for each language 1`] = `
{
  "bladee": "Waylon",
  "promiscuous girl": "promiscue meisje",
}
`;

exports[`toJson returns json with translations for each language 2`] = `
{
  "promiscuous girl": "pasileidusi mergina",
}
`;
